src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp

```diff
@@ -0,0 +1,304 @@
+    }
+}
+
+bool BallPlacementPlayFSM::kickDone(const Update &event)
+{
+    const auto ball_velocity = event.common.world.ball().velocity().length();
+    return (ball_velocity > SHOT_VELOCITY_THRESHOLD_M_PER_S) && !shouldKickOffWall(event);
+}
+
+bool BallPlacementPlayFSM::ballPlaced(const Update &event)
+{
+    Point ball_pos = event.common.world.ball().position();
+    std::optional<Point> placement_point =
+        event.common.world.gameState().getBallPlacementPoint();
+    std::vector<Robot> robots = event.common.world.friendlyTeam().getAllRobots();
+
+    // see if the ball is at the placement destination
+    if (placement_point.has_value())
+    {
+        return comparePoints(ball_pos, placement_point.value(),
+                             PLACEMENT_DIST_THRESHOLD_METERS) &&
+               event.common.world.ball().velocity().length() <
+                   this->ai_config.ai_parameter_config()
+                       .ball_is_kicked_m_per_s_threshold();
+    }
+    else
+    {
+        return true;
+    }
+}
+
+bool BallPlacementPlayFSM::waitDone(const Update &event)
+{
+    std::chrono::time_point<std::chrono::system_clock> current_time =
+        std::chrono::system_clock::now();
+    return static_cast<double>(
+               std::chrono::duration_cast<std::chrono::seconds>(current_time - start_time)
+                   .count()) > BALL_IS_PLACED_WAIT_S;
+}
+
+bool BallPlacementPlayFSM::retreatDone(const Update &event)
+{
+    Point ball_position = event.common.world.ball().position();
+    return distance(ball_position, event.common.world.friendlyTeam()
+                                       .getNearestRobot(ball_position)
+                                       ->position()) > RETREAT_DISTANCE_METERS;
+}
+
+Angle BallPlacementPlayFSM::calculateWallKickoffAngle(const Point &ball_pos,
+                                                      const Rectangle &field_lines)
+{
+    Angle kick_angle;
+    if (ball_pos.x() > field_lines.xMax())
+    {
+        if (ball_pos.y() > 0)
+        {
+            kick_angle = Angle::fromDegrees(45);
+        }
+        else
+        {
+            kick_angle = Angle::fromDegrees(-45);
+        }
+    }
+    else if (ball_pos.x() < field_lines.xMin())
+    {
+        if (ball_pos.y() > 0)
+        {
+            kick_angle = Angle::fromDegrees(135);
+        }
+        else
+        {
+            kick_angle = Angle::fromDegrees(-135);
+        }
+    }
+    else if (ball_pos.y() > field_lines.yMax())
+    {
+        if (ball_pos.x() > 0)
+        {
+            kick_angle = Angle::fromDegrees(135);
+        }
+        else
+        {
+            kick_angle = Angle::fromDegrees(45);
+        }
+    }
+    else if (ball_pos.y() < field_lines.yMin())
+    {
+        if (ball_pos.x() > 0)
+        {
+            kick_angle = Angle::fromDegrees(-135);
+        }
+        else
+        {
+            kick_angle = Angle::fromDegrees(-45);
+        }
+    }
+    return kick_angle;
```

Comment:
Is the first set of if statements checking if the ball is outside of the field lines?! If it isn't, then the kick angle is whatever the default constructor of `Angle` initialise it at (i.e. 0 degrees)?!
Could you maybe add a comment explaining the logic here? Is the idea to try to kick somewhat towards the middle of the field?

Reply:
we only perform the kick of the ball is outside the field, so it's some increment of 45 degrees. I'm not sure what you mean by "0 degrees" here

Reply:
I guess I was thinking of the case where we don't go into any of the if statements, but I guess that shouldn't happen because of the `shouldKickOffWall` guard.